Use Pydantic for generating the data model (#2211)
ICTU/quality-time: components/server/README.md

## Data model

> The data model package describes the domain model used by the application. It allows for a frontend that doesn't need to know about specific metrics and sources. On server start up, it checks whether the data model has changed, and if so, imports it into the database.

Reviewer: The package is currently called "data" instead of "data model" since it also still contains the example reports. Maybe it would be a good idea to give the data model it's own package and make call the data package example_data?

The data model package consists of a meta model and the data model itself. The data model consists of four major parts:

- Scales
- Metrics
- Sources
- Subjects

The meta model uses Pydantic to specify the components and attributes of the data model.

### Scales

> The `scales` part of the data model defines the scales uses. At the time of writing these include an absolute count scale, a percentage scale, and a version number scale.

Reviewer: ```suggestion
The `scales` part of the data model defines the scales that can be used to measure metrics. At the time of writing these include an absolute count scale, a percentage scale, and a version number scale.
```

Each metric defines the scales it supports.

### Metrics

> The `metrics` part of the data model is an object where the keys are the metric types and the values are objects describing the metric type. Metrics have the following meta model:

Reviewer: ```suggestion
The `metrics` part of the data model is a dictionary with all supported metric types. The keys are the metric type names and the values are objects describing the metric type. All metrics types have the following meta model:
```

    near_target: str = "10"
    sources: list[str] = Field(..., min_items=1)
    default_source: Optional[str] = None
    tags: list[Tag] = []
```

The `name` is the default name of metrics of this type. The `description` describes what the metric measures. These are part of the `DescribedModel`.

The `scales` list shows which scales the metric supports and the `default_scale` specifies which scale is the default scale. 

The `unit` is the default unit of the metric, e.g. lines of code, security warnings, or in the above example, complex units. 

The `addition` determines how values from multiple sources are combined: possible values are `max`, `min`, and `sum`. 

The `direction` specifies whether smaller measurement values are better or worse.

The `target` is the default target value for the metric. The `near_target` is when the metric becomes red. Values between `target` and `near_target` are yellow.

The list of `sources` contains the keys of source types that support this metric type.

Finally, `tags` are strings used to group related metrics.

Users with sufficient rights can override the default name, unit, and target of metrics via the user interface.

### Sources

> The `sources` part of the data model is an object where the keys are the source types and the values are objects describing the source. Sources have the following meta model:

Reviewer: ```suggestion
The `sources` part of the data model is a dictionary that describes all supported source types. The keys are the source type names and the values are objects describing the source type. All source types have the following meta model:
```
